### PR TITLE
[feature] Add a QgsScaleCalculator function to calculate canvas width from a given extent, scale, and dpi

### PR DESCRIPTION
**Pull request by nirvn**
## Description

I've had the need for this one time too many, time to add it to the API :)

A new calculateCanvasWidth( extent, scale ) was added to the QgsScaleCalculator class to derive canvas width in pixel unit when providing an extent, a scale, and a DPI.

### CODE REVIEW

File: python/core/auto_generated/qgsscalecalculator.sip.in [hunk 1]
```diff
@@ -65,6 +65,16 @@ Calculate the scale denominator
 :param canvasWidth: Width of the map canvas in pixel (physical) units
 
 :return: scale denominator of current map view, e.g. 1000.0 for a 1:1000 map.
+%End
+
+    double calculateCanvasWidth( const QgsRectangle &mapExtent, double scale ) const;
```
Comment (m-kuhn) [inline python/core/auto_generated/qgsscalecalculator.sip.in:70]: I think we could do better with units and terms.
1. we are mixing rectangle/size with double/length
2. this works for with or height, so the term `width` shouldn't be encoded into the naming

Option a:

```suggestion
    QSizeF calculateCanvasSize( const QgsRectangle &mapExtent, double scale ) const;
```

Option b:

```suggestion
    // Usually pass in `mapExtent.width()` for your requirements
    double canvasLengthToPixelLength( double canvasLength, double scale ) const;
```

Comment (nirvn) [reply python/core/auto_generated/qgsscalecalculator.sip.in:70]: I love option A.

Comment (nyalldawson) [reply python/core/auto_generated/qgsscalecalculator.sip.in:70]: Can we replace "canvas" with "image"?

File: src/core/qgsscalecalculator.h [hunk 2]
```diff
@@ -72,6 +72,14 @@ class CORE_EXPORT QgsScaleCalculator
      */
     double calculate( const QgsRectangle &mapExtent, double canvasWidth ) const;
 
+    /**
+     * Calculate the canvas width in pixel (physical) units
+     * \param mapExtent QgsRectangle containing the current map extent
+     * \param scale Scale denominator, e.g. 1000.0 for a 1:1000 map
+     * \returns canvas width in pixel
+     */
```
Comment (nyalldawson) [inline src/core/qgsscalecalculator.h:80]: Needs "since"

File: python/core/auto_generated/qgsscalecalculator.sip.in (hunk 1 above)
Comment (nirvn) [reply python/core/auto_generated/qgsscalecalculator.sip.in:70]: Done.

**Comment by nyalldawson** [comment]
Got some tests coming?

**Comment by nirvn** [comment]
@nyalldawson , ya.

**Comment by nirvn** [comment]
@nyalldawson , test added.

File: tests/src/python/test_qgsscalecalculator.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+from qgis.PyQt.QtCore import Qt, QSizeF
+from qgis.PyQt.QtTest import QSignalSpy
+
+from qgis.core import QgsScaleCalculator, QgsRectangle, QgsUnitTypes
+from qgis.testing import start_app, unittest
+
+start_app()
+
+
+class TestQgsScaleWidget(unittest.TestCase):
```
Comment (nyalldawson) [inline tests/src/python/test_qgsscalecalculator.py:24]:
```suggestion
class TestQgsScaleCalculator(unittest.TestCase):
```

**Comment by nirvn** [comment]
@nyalldawson , suggestion applied.

**Review by m-kuhn — APPROVED** [review]

**Pull request merged**